grafanaplane/oncall/escalation.libsonnet

```diff
@@ -0,0 +1,140 @@
+local d = import 'github.com/jsonnet-libs/docsonnet/doc-util/main.libsonnet';
+local xtd = import 'github.com/jsonnet-libs/xtd/main.libsonnet';
+
+local raw = import '../zz/main.libsonnet';
+
+{
+  local this = self,
+  local escalation = raw.oncall.v1alpha1.escalation,
+
+  '#new': d.func.new(
+    |||
+      `new` creates an Escalation. `escalationChainName` is the resource name
+      (`myescalationchain.metadata.name`) of the parent Escalation Chain.
+      `position` is the position of the Escalation in the chain. `type` is the
+      Escalation type. `providerName` is the resource name
+      (`myprovider.metadata.name`) of the Provider. `resourceId` defaults to
+      `<escalationChainName>-<position>`.
+    |||,
+    [
+      d.argument.new('escalationChainName', d.T.string),
+      d.argument.new('position', d.T.number),
+      d.argument.new('type', d.T.string),
+      d.argument.new('providerName', d.T.string),
+      d.argument.new(
+        'resourceId',
+        d.T.string,
+        default='<escalationChainName>-<position>'
+      ),
+    ]
+  ),
+  new(
+    escalationChainName,
+    position,
+    type,
+    providerName,
+    resourceId='%s-%d' % [escalationChainName, position]
+  )::
+    escalation.new(resourceId)
+    + escalation.spec.parameters.providerConfigRef.withName(providerName)
+    + escalation.spec.parameters.forProvider.escalationChainRef.withName(escalationChainName)
+    + escalation.spec.parameters.forProvider.withPosition(position)
+    + escalation.spec.parameters.forProvider.withType(type),
+
+  '#newNotifyOnCallFromSchedule': d.func.new(
```

Comment:
I see many constructors like this (`new+<Type>`) with lots of repetition, I feel we can do better.

```jsonnet
{
  escalation: {
    new(escalationChainName, position):
      escalation.new('%s-%d' % [escalationChainName, position])
      + escalation.spec.parameters.forProvider.escalationChainRef.withName(escalationChainName)
      + escalation.spec.parameters.forProvider.withPosition(position),

    withName(name):
      escalation.metadata.withName(name),

    withNotifyOnCallFromSchedule(scheduleName):
      escalation.spec.parameters.forProvider.withType('notify_on_call_from_schedule')
      + escalation.spec.parameters.forProvider.withNotifyOnCallFromScheduleRef(
        escalation.spec.parameters.forProvider.notifyOnCallFromScheduleRef.withName(scheduleName)
      ),

    withNotifyPersons(persons):
      escalation.spec.parameters.forProvider.withType('notify_persons')
      + escalation.spec.parameters.forProvider.withPersonsToNotify(persons),
  },
}
```

This'll detach the type from the position parameter, that'll allow the end user to modify the flow without having to fiddle with the numbers:

```jsonnet
local chainName = 'myChain';
[
  escalation.new(chainName, 1)
  + withNotifyPersons(['sasha@grafana.com']),
  escalation.new(chainName, 2)
  + withNotifyOnCallFromSchedule('toSchedule'),
]
```

We can even take it one step further and infer the chain and position while creating the chain:

```jsonnet
escalationChain.new('myChain')
+ escalationChain.withSteps([
  step.notifyPersons(['sasha@grafana.com']),
  step.notifyOnCallFromSchedule('toSchedule'),
])
```

In this case the code would completely different, something like this:

```jsonnet
{
  escalationChain: {
    new(name): {
      chainName:: name,
      chain:
        escalationChain.new(xtd.ascii.stringToRFC1123(name))
        + escalationChain.spec.parameters.forProvider.withName(name),
    },

    withSteps(steps): {
      steps:
        std.mapWithIndex(
          function(position, step)
            escalation.new(xtd.ascii.stringToRFC1123('%s-%d' % [escalationChainName, position]))
            + escalation.spec.parameters.forProvider.escalationChainRef.withName(self.chainName)
            + escalation.spec.parameters.forProvider.withPosition(position)
            + step,
          steps
        ),
    },

    step: {
      notifyOnCallFromSchedule(scheduleName):
        escalation.spec.parameters.forProvider.withType('notify_on_call_from_schedule')
        + escalation.spec.parameters.forProvider.withNotifyOnCallFromScheduleRef(
          escalation.spec.parameters.forProvider.notifyOnCallFromScheduleRef.withName(scheduleName)
        ),

      notifyPersons(persons):
        escalation.spec.parameters.forProvider.withType('notify_persons')
        + escalation.spec.parameters.forProvider.withPersonsToNotify(persons),
    },
  },
}
```

I didn't test any of this, so not sure if it all works like I'm intending.